src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+            .send()
+            .await
+            .map_err(|err| {
+                CredentialsError::from_str(false, format!("failed to request token: {}", err))
+            })?;
+
+        let status = res.status();
+        println!("[execute] status: {:?}", status);
+        println!("[execute] response: {:?}", res);
+        if !status.is_success() {
+            return Err(CredentialsError::from_str(
+                false,
+                format!("error exchanging token, failed with status {status}"),
+            ));
+        }
+        let token_res = res
+            .json::<TokenResponse>()
+            .await
+            .map_err(|err| CredentialsError::new(false, err))?;
+        Ok(token_res)
+    }
+}
+
+/// TokenResponse is used to decode the remote server response during
+/// an oauth2 token exchange.
+#[derive(Deserialize, Default, PartialEq, Debug)]
+pub struct TokenResponse {
+    pub access_token: String,
+    pub issued_token_type: String,
+    pub token_type: String,
+    pub expires_in: u64,
+    pub scope: String,
+    pub refresh_token: Option<String>,
+}
+
+/// Authentication style via headers or form params.
+/// See https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub enum ClientAuthStyle {
+    InParams,
+    InHeader,
+}
+
+/// ClientAuthentication represents an OAuth client ID and secret and the
+/// mechanism for passing these credentials as stated
+/// in https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1.
+#[derive(Clone, Debug)]
+pub struct ClientAuthentication {
+    pub auth_style: ClientAuthStyle,
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+}
+
+impl Default for ClientAuthentication {
+    fn default() -> Self {
+        Self {
+            auth_style: ClientAuthStyle::InParams,
+            client_id: None,
+            client_secret: None,
+        }
+    }
+}
+
+impl ClientAuthentication {
+    // Add authentication to a Secure Token Service exchange request.
+    // Modifies either the passed headers or form parameters
+    // depending on the desired authentication format.
+    #[allow(dead_code)]
+    pub fn inject_auth(&self, headers: &mut http::HeaderMap, params: &mut HashMap<&str, String>) {
+        if let (Some(client_id), Some(client_secret)) =
+            (self.client_id.clone(), self.client_secret.clone())
+        {
+            match self.auth_style {
+                ClientAuthStyle::InHeader => {
+                    let plain_header = format!("{}:{}", client_id, client_secret);
+                    let encoded =
+                        base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(plain_header);
+                    let header = http::HeaderValue::from_str(format!("Basic {encoded}").as_str());
+                    if let Ok(value) = header {
+                        headers.insert("Authorization", value);
+                    }
+                }
+                _ => {
```

Comment:
Be explicit, use `ClientAuthStyle::InParams` that way you will get an error if there is ever another style.